### PR TITLE
feat(edgeless): shape tool drag and animation

### LINKED ISSUE #4538: Edgeless Toolbar optimization
shot-2023-09-01-18 42 41@2x

shot-2023-09-01-19 55 20@2x

[the design file link](https://www.figma.com/file/UTTtfVCcStw1dKUg0gCSPo/%F0%9F%9B%9E-Design-System?type=design&node-id=4454%3A31690&mode=design&t=2HhMqxlgFat698b8-1)

- [x] Modify the style of the pen and eraser. Keep the hover elongation effect.
- [x] Modify the style of the connector to be a separate icon.
- [x] Display icons and badges on the left side, but not on the right side.
- [ ] Enlarge the shape of the Shape, supporting dragging and animations.
- [x] The pen/shape/text supports color changes.
- [x] Secondary menus support scrolling, and the left and right buttons enlarge on hover.

### PR DESCRIPTION
**Pull request by lzlme**
close #4538 

https://github.com/toeverything/blocksuite/assets/117659326/57a29f94-9b32-4539-ad9e-b0379d5e5abd

### CODE REVIEW

**Comment by hwangdev97** [comment]
Good job 👍 

There are still some details: 

1. In the click-activated scenario, it is necessary to deal with the problem of super small shapes being created due to slight movement.

I think it would be better to start creating only when the displacement is greater than a certain number, maybe 20px or 50px, 
Depends on specific implementation

shot-2023-09-22-13 57 59@2x

2. The placement of the shape has shifted, which is not quite what we expected. A slight displacement will have the metaphor of letting go, but the current displacement is too large, which is wrong. 

My suggestion is to position it about 4px lower and right than the placeholder.


https://github.com/toeverything/blocksuite/assets/24713927/cb2c7cc2-c293-4737-9b2d-b8323b38628e


3. This bottom shape area needs to be larger

Current preview's width is `92px` design file is `102px`. And Inactive shapes can be moved further down.

shot-2023-09-22-14 08 43@2x

File: packages/blocks/src/page-block/edgeless/components/toolbar/shape/shape-element.ts [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+  @state()
+  private startCoord: Coord = { x: 0, y: 0 };
+
+  @state()
+  private dragging: boolean = false;
+
+  @state()
+  private isOutside: boolean = false;
+
+  private onDragStart = (coord: Coord) => {
```
Comment (doodlewind) [inline packages/blocks/src/page-block/edgeless/components/toolbar/shape/shape-element.ts:117]: Please add  `_` prefix for private methods and variables.

File: packages/blocks/src/page-block/edgeless/components/toolbar/shape/shape-element.ts [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+interface TransformMap {
+  [key: string]: {
+    x: number;
+    y: number;
+    scale: number;
+    origin: string;
+  };
+}
+
+@customElement('edgeless-shape-element')
```
Comment (doodlewind) [inline packages/blocks/src/page-block/edgeless/components/toolbar/shape/shape-element.ts:47]: This naming seems to be overkill.

File: packages/blocks/src/page-block/edgeless/components/toolbar/shape/shape-element.ts (hunk 1 above)
Comment (lzlme) [reply packages/blocks/src/page-block/edgeless/components/toolbar/shape/shape-element.ts]: done

**Review by hwangdev97 — APPROVED** [review]

**Review by donteatfriedrice — APPROVED** [review]

**Review by doodlewind — APPROVED** [review]
Thanks for polishing this!

**Pull request merged**